src/main/java/net/minecraftforge/server/permission/handler/IPermissionHandler.java

```diff
@@ -0,0 +1,90 @@
+/*
+ * Minecraft Forge
+ * Copyright (c) 2016-2021.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation version 2.1
+ * of the License.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+
+package net.minecraftforge.server.permission.handler;
+
+import net.minecraft.resources.ResourceLocation;
+import net.minecraft.server.level.ServerPlayer;
+import net.minecraft.util.StringRepresentable;
+import net.minecraftforge.server.permission.PermissionAPI;
+import net.minecraftforge.server.permission.events.PermissionGatherEvent;
+import net.minecraftforge.server.permission.nodes.PermissionDynamicContext;
+import net.minecraftforge.server.permission.nodes.PermissionNode;
+
+import java.util.Set;
+import java.util.UUID;
+
+/**
+ * This is the Heart of the PermissionAPI, it manages {@link PermissionNode}s
+ * as well as it handles all permission queries.
+ *
+ * <p>Note: You do not need to implement a PermissionHandler to query for permissions.</p>
+ *
+ * @implNote The {@link DefaultPermissionHandler} does forward all permission queries to the PermissionNodes default resolver.
+ * @apiNote You can implement your own PermissionHandler using the {@link PermissionGatherEvent.Handler} event.
+ */
+public interface IPermissionHandler
+{
+    /**
+     * {@return an identifier for the PermissionHandler}
+     */
+    ResourceLocation getIdentifier();
+
+    /**
+     * <strong>Mods must use {@link PermissionGatherEvent.Nodes}</strong>
+     *
+     * <p>It is guaranteed that all permission nodes passed in here have a unique name.</p>
+     *
+     * @param node PermissionNode to be registered
+     */
+    void registerNode(PermissionNode<?> node);
```

Comment:
Should this instead be a factory method so that people cant some in at runtime and add random nodes?

Reply:
I think that's unnecessary, the final active PermissionHandler isn't accessible.
So registering new nodes outside the events isn't really possible.
Additionally, this would require mods to reconstruct their PermissionNodes whenever a singleplayer world is loaded.
Which I fear might tempt people to return to the old way of creating dynamic permissions, 
rather than using the new more universal approach.